From bef94704602b617eacc68540080eaec212aad01e Mon Sep 17 00:00:00 2001
From: Guo Jie <guojie@loongson.cn>
Date: Tue, 27 Jun 2023 10:03:53 +0800
Subject: [PATCH 03/13] gcc-8.3.0: Fix the bug in loongarch_emit_stack_tie.

Which may result in implicit references to $fp when frame_pointer_needed is false,
causing regs_ever_live[$fp] to be true when $fp is not explicitly used,
resulting in $fp being used as the target replacement register in the rnreg pass.

The bug originates from SPEC2017 541.leela_r(-flto).

Change-Id: I11316853e230845cefe6f54189f616432b80cca1
---
 src/gcc/config/loongarch/loongarch.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/gcc/config/loongarch/loongarch.c b/src/gcc/config/loongarch/loongarch.c
index 69043f363..176b7933a 100644
--- a/src/gcc/config/loongarch/loongarch.c
+++ b/src/gcc/config/loongarch/loongarch.c
@@ -1193,7 +1193,9 @@ static void
 loongarch_emit_stack_tie (void)
 {
   emit_insn (PMODE_INSN (gen_stack_tie,
-			 (stack_pointer_rtx, hard_frame_pointer_rtx)));
+			 (stack_pointer_rtx, frame_pointer_needed
+					     ? hard_frame_pointer_rtx
+					     : stack_pointer_rtx)));
 }
 
 #define PROBE_INTERVAL (1 << STACK_CHECK_PROBE_INTERVAL_EXP)
-- 
2.31.1

